R/meta.R

```diff
@@ -0,0 +1,108 @@
+
+#  Fixed-effects meta-analysis
+#  @param beta 
+# @param se
+# importFrom("stats", "pchisq")
+
+meta.F <- function(beta, se){
+  #returns inverse-variance weighted meta-analysis estimate, SE and P-value.
+  beta.F = sum(beta / se^2) / sum(1 / se^2)
+  se.F = 1 / sqrt(sum(1 / se^2))
+  p.F = pchisq( (beta.F / se.F)^2, df = 1, lower.tail = F)
+  return(c(beta = beta.F, se = se.F, p = p.F))
+}
+
+
+meta_analyse <- function(dataset, method = "fixed"){
+length_dt <- getLength(dataset)
+nsnps <- dim(getData(dataset,1))[1]
+beta <- matrix(ncol = length_dt, nrow = nsnps)
+se <- matrix(ncol = length_dt, nrow = nsnps)
+
+for (i in 1:length_dt){
+beta[,i] <- getData(dataset,i)$beta
+se[,i] <- getData(dataset,i)$se
+}
+
+if(method == "fixed"){
+  meta <- matrix(ncol = 3, nrow = nsnps)
+  meta <- sapply(1:nsnps, function(i) {
+  meta[i,] <- meta.F(beta[i,],se[i,])})
+}
+
+return(as_tibble(t(meta)))
+}
+
+#' Meta analysis
+#' @description Statistical combination of the
+#' results from two or more separate studies. 
+#' It uses the fixed-effect model assuming that one true 
+#' effect size underlies all the studies in the meta-analysis. 
+#' @param dataset gwasglue2 DataSet object
+#' @param method Uses fixed-effect model. Default ('"fixed"')
+#' @importFrom stats pchisq
+#' @return gwasglue2 SummarySet object
+meta_analysis <- function(dataset, method = "fixed") {
+
+  length_dt <- getLength(dataset)
+  nsnps <- dim(getData(dataset,1))[1]
+
+  lapply(1:length_dt, function(i) {
+    if (is.null(getMetadata(getSummarySet(dataset,i))$sample_size) || is.na(getMetadata(getSummarySet(dataset,i))$sample_size)){
+      stop("No sample size information in metadata for at least one of the SummarySets. More details on how to add to metadata in 'help(gwasglue2::create_metadata)' and 'help(gwasglue2::getMetadata)'." )
+    }
+    })
+
+
+  # meta data id
+  ids <- unlist(lapply(1:length_dt, function(i) {
+    id <- getMetadata(getSummarySet(dataset,i))$id }))
+  id <- paste(ids, collapse = "||")
+
+  # meta data sample size
+  n <- lapply(1:length_dt, function(i) {
+    n <- getMetadata(getSummarySet(dataset,i))$sample_size})
+  n_meta <- sum(unlist(n))
+
+  # meta data trait
+  traits <- unlist(lapply(1:length_dt, function(i) {
+    t <- getMetadata(getSummarySet(dataset,i))$trait}))
+  trait <- paste(unique(traits), collapse = "||")
+
+  # eaf weighted mean
+  eaf <- lapply(1:length_dt, function(i) {
+    m <- getData(dataset,i)$eaf * getMetadata(getSummarySet(dataset,i))$sample_size
+    })
+  eaf_mean <- Reduce("+", eaf) / n_meta
+
+  # create metadata object
+  metadata <- create_metadata(metadata = NULL,
```

Comment:
This assumes that those values are all present, but e.g. for this set of traits `chd_id <- c("ieu-a-7",  "ukb-d-I9_IHD", "finn-b-I9_CHD")` the last one has no sample size but it does have ncase and ncontrol, and so it errors. So the sample size is a sum of those two, but really the case control numbers are important too. Not completely sure what to do with this but perhaps at least would be helpful to be able to easily add sample size (or any metadata value) using a setter function

Reply:
When creating metadata for each summaryset the user can add the `sample_size`, and if none is given the `create_summaryset()` function tries to fill it using the `max($n)` from the summary statistics. In `meta.R`, there is an error if the `sample_size` is missing from one of the summarysets and advises to fill that value using `create_metadata()`. I could create a function like `add_to_metadata`?

Reply:
I changed the `SetMetadata()` and now we can easyly add to the Metadata